STYLE_GUIDE.md

```diff
@@ -0,0 +1,461 @@
+# Pony Standard Library Style Guide
+
+This is a document describing the preferred practices for writing code for the Pony standard library. Other projects are free to use the same style, but you are free to use any style in other projects. The following conventions are meant to ensure consistency and readability of Pony code. Any of the following rules may be broken in situations where the recommended practice would diminish readability.
+
+
+## Formatting
+
+### Line Length
+
+The maximum line length is 80 columns.
+
+### Trailing Whitespace
+
+No trailing whitespace at the end of lines.
+
+### Indentation
+
+Use 2 spaces for indentation, No hard tabs.
+
+```pony
+actor Main
+  new create(env: Env) =>
+    env.out.print("Hello, world!")
+```
+
+### Blank Lines
+
+- One blank line is placed between top-level declarations such as classes, primitives, actors, etc.
+- There is no blank lines between the class declaration and its first member or documentation.
+- Fields have no blank lines between them.
+- Functions are separated from other members by one line.
+
+  ```pony
+  // OK
+  class Foo
+  let bar: Bar
+  let baz: Baz
+
+  fun foo() =>
+  // ...
+
+  fun bar() =>
+  // ...
+
+  // OK
+  class Foo
+  """
+  Some documentation
+  """
+  let bar: Bar
+
+  fun foo() =>
+  // ...
+
+  // OK
+  class Foo
+  fun foo() =>
+  // ...
+
+  // Not OK
+  class Foo
+
+  let bar: Bar
+  let baz: Baz
+  fun foo() =>
+  // ...
+  ```
+
+- Blank lines between __one-line__ functions or primitives may be omitted if they are related.
+
+  ```pony
+  primitive Red    fun apply(): U32 => 0xFF0000FF
+  primitive Green  fun apply(): U32 => 0x00FF00FF
+  primitive Blue   fun apply(): U32 => 0x0000FFFF
+
+  type Color is (Red | Green | Blue)
+  ```
+
+### Whitespace
+
+- Use spaces around binary operators and after commas, colons, and semicolons. Do not put spaces around brackets or parentheses.
+
+  ```pony
+  let x: USize = 3
+
+  x = x + (5 * x)
+
+  (let a, let b) = ("a", "b")
+
+  let ns = [as USize: 1; 2; 3]
+  ```
+
+- The `not` operator is surrounded by spaces, but the `-` operator is not followed by a space.
+
+  ```pony
+  if not x then -a end
+  ```
+
+- Lambda expressions follow the rules above except that a space is placed before the closing brace.
+
+  ```pony
+  // OK
+  {(a: USize, b: USize): USize => a + b }
+
+  // OK
+  {(a: USize, b: USize): USize =>
+    a + b }
+
+  // Not OK
+  {(a: USize, b: USize): USize => a + b}
+
+  // Not OK
+  { (a: USize, b: USize): USize => a + b }
+
+  // Not OK
+  {(a: USize, b: USize): USize =>
+    a + b
+  }
```

Comment:
I think I disagree about that - I don't think a lambda is in the same category as the other two.

In part because it will be quite common for lambdas to become rather complex by comparison, with other levels of nested indentation. Arrays can technically contain arbitrarily complex blocks of code, but in practice I think any nested levels of indentation would be quite rare.

Here's an example of a lambda that isn't really that complex conceptually, but due to the nested levels of indentation, using the "same line" closing bracket strategy would look quite awkward:

```pony
rs.next({(p: zmq.SocketPeer, m: zmq.Message)(rs) =>
  if m.size() > 0
    peers = state.peers()
    for peer in peers.values()
      peer.send(recover zmq.Message.>append(m).>push("S") end)
    end
  end
}
```

Just imagine how it would look with a large `match` block thrown in.